### PR TITLE
Detach fake tensors into val, so they aren't affected by metadata mutation

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #89140

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

### CODE REVIEW

File: torch/fx/experimental/proxy_tensor.py [hunk 1]
```diff
@@ -108,13 +108,13 @@ def has_proxy(obj):
 
 def set_meta(proxy, val):
     if isinstance(val, FakeTensor):
-        proxy.node.meta['val'] = val
+        proxy.node.meta['val'] = val.detach()
```
Comment (ezyang) [inline torch/fx/experimental/proxy_tensor.py:111]: What if val is a constant

Comment (ezyang) [reply torch/fx/experimental/proxy_tensor.py:111]: @SherlockNoMad all of the requires_grad is False for all the val in the graph

Comment (ezyang) [reply torch/fx/experimental/proxy_tensor.py:111]: Cannot use val to drive interpreter anymore.

**Review by bdhirsh — APPROVED** [review]
Are there any docs/comments for `meta['val']`? We might want to write out what the invariants are (accurate metadata except for autograd meta)

File: torch/fx/experimental/proxy_tensor.py (hunk 1 above)
Comment (ezyang) [reply torch/fx/experimental/proxy_tensor.py:111]: So it turns out that it is pretty difficult to get accurate `requires_grad` metadata on the fake tensors if you detach. This is because, at the point proxy tensor is running (under autograd), we don't actually have accurate requires grad metadata yet! It gets set correctly on the tensor when we return from proxy tensor. If we save the fake tensor at this point in time, we don't get the correct setting reflected on the fake tensor when it is modified in place.

Comment (ezyang) [reply torch/fx/experimental/proxy_tensor.py:111]: So I'm going to skip requires_grad recording and file a bug for it for later, if anyone cares.
